### PR TITLE
feat(ext/http): Automatic compression for Deno.serve

### PR DESCRIPTION
**Pull request by mmastrac**
`Content-Encoding: gzip` support for `Deno.serve`. This doesn't support Brotli (`br`) yet, however it should not be difficult to add. Heuristics for compression are modelled after those in `Deno.serveHttp`.

Tests are provided to ensure that the gzip compression is correct. We chunk a number of different streams (zeros, hard-to-compress data, already-gzipped data) in a number of different ways (regular, random, large/small, small/large).

### CODE REVIEW

File: ext/http/http_next.rs [hunk 1]
```diff
@@ -475,6 +485,131 @@ pub fn op_set_response_headers(
...
+fn ensure_vary_accept_encoding(hmap: &mut HeaderMap) {
+  if let Some(v) = hmap.get_mut(hyper::header::VARY) {
+    if let Ok(s) = v.to_str() {
+      if !s.to_lowercase().contains("accept-encoding") {
+        *v = format!("Accept-Encoding, {s}").try_into().unwrap()
+      }
+      return;
+    }
+  }
+  hmap.insert(
```
[the remark this answers is not in the document]
Comment (mmastrac) [reply ext/http/http_next.rs:588]: I double-checked this API and it's somewhat poorly named -- the operation is really "insert, replace if already exists". 

https://docs.rs/http/latest/http/header/struct.HeaderMap.html#method.insert

File: ext/http/response_body.rs [hunk 2]
```diff
@@ -17,9 +21,44 @@ use deno_core::CancelTryFuture;
...
+  /// not register a waker and should be called again at the lowest level of this code. Generally this
+  /// will only be returned from compression streams that require additional buffering.
+  NoData,
+  /// Stream provided trailers.
+  // TODO(mmastrac): We are threading trailers through the response system to eventually support Grpc.
+  #[allow(unused)]
+  Trailers(HeaderMap),
+  /// Stream failed.
+  Error(AnyError),
+}
```
Comment (bartlomieju) [inline ext/http/response_body.rs:48]: This is a really good abstraction

File: ext/http/response_body.rs [hunk 3]
```diff
@@ -62,19 +101,41 @@ impl Future for CompletionHandle {
...
+
+#[derive(PartialEq, Eq)]
+pub enum Compression {
+  None,
+  GZip,
+}
+
+pub enum ResponseStream {
+  /// A resource stream, piped in fast mode.
```
Comment (bartlomieju) [inline ext/http/response_body.rs:120]: Add a comment that this kind doesn't involve going to JS to obtain more data and always stays in Rust while streaming

File: ext/http/response_body.rs [hunk 4]
```diff
@@ -206,6 +290,243 @@ impl Drop for ResponseBytes {
...
+      Err(err) => ResponseStreamResult::Error(err),
+      Ok(buf) => {
+        if buf.is_empty() {
+          if self.auto_close {
+            self.stm.clone().close();
+          }
+          ResponseStreamResult::EndOfStream
+        } else {
+          // Re-arm the future
+          self.future = self.stm.clone().read(64 * 1024);
```
Comment (bartlomieju) [inline ext/http/response_body.rs:344]: Is this clone expensive?

Comment (lucacasonato) [reply ext/http/response_body.rs:344]: No, it's an RC.

File: cli/tests/unit/serve_test.ts [hunk 5]
```diff
@@ -1425,41 +1425,6 @@ Deno.test(
   },
 );
 
-Deno.test(
-  { permissions: { net: true, write: true, read: true } },
-  async function httpServerCorrectSizeResponse() {
```
[the remark this answers is not in the document]
Comment (mmastrac) [reply cli/tests/unit/serve_test.ts:1430]: This one was combined into the combinatorial test that I added

File: cli/tests/unit/serve_test.ts [hunk 6]
```diff
@@ -2012,38 +1977,146 @@ Deno.test(
   },
 );
 
-Deno.test(
-  { permissions: { net: true, write: true, read: true } },
-  async function httpServerSendFile() {
```
[the remark this answers is not in the document]
Comment (mmastrac) [reply cli/tests/unit/serve_test.ts:2017]: This one was also combined into the combinatorial test that I added

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**